File: build_lexicon.cpp
#include <iostream>
#include <fstream>
#include <filesystem>
#include <map>
#include <vector>
#include <sstream>
#include <algorithm>

using namespace std;
namespace fs = std::filesystem;

string OUTPUT_PATH = "D:/DSA-Project/search_engine/backend/app/indexing/output/";
string DOCS_PATH = "D:/Dataset/Cleaned1/";

vector<string> tokenize(const string &text) {
    vector<string> tokens;
    string word;

    for (char c : text) {
        if (isalnum(c)) word += tolower(c);
        else {
            if (!word.empty()) {
                tokens.push_back(word);
                word.clear();
            }
        }
    }
    if (!word.empty()) tokens.push_back(word);

    return tokens;
}

int main() {
    map<string, int> lexicon;
    int wordID = 1;

    for (const auto &file : fs::directory_iterator(DOCS_PATH)) {
        if (file.path().extension() == ".txt") {
            ifstream in(file.path());
            stringstream buffer;
            buffer << in.rdbuf();
            string text = buffer.str();
            vector<string> tokens = tokenize(text);

            for (auto &t : tokens) {
                if (lexicon.find(t) == lexicon.end()) {
                    lexicon[t] = wordID++;
                }
            }
        }
    }

    ofstream out(OUTPUT_PATH + "lexicon.csv");
    out << "word,wordID\n";
    for (auto &p : lexicon) {
        out << p.first << "," << p.second << "\n";
    }

    cout << "lexicon.csv generated!" << endl;
    return 0;
}

File: build_lexicon.py
import os, json, gzip, re

JSON_PATH = "D:/DATA"
OUTPUT = "D:/DSA-Project/search_engine/backend/app/indexing/output/lexicon.csv"

def tokenize(text):
    return re.findall(r"[a-z0-9]+", text.lower())

lexicon = {}
word_id = 1

def process_file(path):
    global word_id
    if path.endswith(".gz"):
        f = gzip.open(path, "rt", encoding="utf-8")
    else:
        f = open(path, "r", encoding="utf-8")

    data = json.load(f)

    text = ""
    if "abstract" in data:
        for x in data["abstract"]:
            text += x.get("text", "") + " "
    if "body_text" in data:
        for x in data["body_text"]:
            text += x.get("text", "") + " "

    for token in tokenize(text):
        if token not in lexicon:
            lexicon[token] = word_id
            word_id += 1

for file in os.listdir(JSON_PATH):
    process_file(os.path.join(JSON_PATH, file))

with open(OUTPUT, "w", encoding="utf-8") as out:
    out.write("word,wordID\n")
    for w, i in lexicon.items():
        out.write(f"{w},{i}\n")

print("Lexicon created")
